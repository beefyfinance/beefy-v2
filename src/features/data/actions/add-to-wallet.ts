import type { ChainEntity } from '../entities/chain';
import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { TokenEntity } from '../entities/token';
import { selectTokenByAddress } from '../selectors/tokens';
import { getSingleAssetSrc } from '../../../helpers/singleAssetSrc';

export type AddToWalletParams = {
  chainId: ChainEntity['id'];
  tokenAddress: TokenEntity['address'];
  customIconUrl?: string;
};

export type AddTokenToWalletPayload = {
  token: TokenEntity;
  iconUrl: string;
};

function getTokenIconUrl(
  id: TokenEntity['id'],
  chainId: TokenEntity['chainId'],
  customUrl?: string
): string {
  if (customUrl) {
    return customUrl;
  }

  const assetSrc = getSingleAssetSrc(id, chainId);
  if (assetSrc) {
    return `${window.location.origin}${assetSrc}`;
  }

  return '';
}

export const addTokenToWalletAction = createAsyncThunk<
  AddTokenToWalletPayload,
  AddToWalletParams,
  { state: BeefyState }
>('addToWallet/open', async ({ chainId, tokenAddress, customIconUrl }, { getState }) => {
  const state = getState();
  const token = selectTokenByAddress(state, chainId, tokenAddress);
  const iconUrl = getTokenIconUrl(token.id, token.chainId, customIconUrl);

  return {
    token,
    iconUrl,
  };
});
