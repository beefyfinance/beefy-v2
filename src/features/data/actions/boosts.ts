import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { TokenAllowance } from '../apis/allowance/allowance-types';
import type { FetchAllBalancesResult } from '../apis/balance/balance-types';
import { getAllowanceApi, getBalanceApi, getConfigApi } from '../apis/instances';
import type { BoostEntity } from '../entities/boost';
import type { ChainEntity } from '../entities/chain';
import { selectBoostById } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectErc20TokenByAddress } from '../selectors/tokens';
import { selectVaultById } from '../selectors/vaults';
import type { BoostCampaignConfig, BoostConfig, BoostPartnerConfig } from '../apis/config-types';

// given the list of vaults is pulled from some api at some point
// we use the api to create an action
// this action should return just enough data for the state to work with

export interface FulfilledAllBoostsPayload {
  boostsByChainId: Record<ChainEntity['id'], BoostConfig[]>;
  partnersById: Record<string, BoostPartnerConfig>;
  campaignsById: Record<string, BoostCampaignConfig>;
}

export const fetchAllBoosts = createAsyncThunk<FulfilledAllBoostsPayload>(
  'boosts/fetchAllBoosts',
  async () => {
    const api = await getConfigApi();
    return api.fetchAllBoosts();
  }
);

interface InitBoostFormParams {
  boostId: BoostEntity['id'];
  mode: 'stake' | 'unstake';
  walletAddress: string | undefined;
}

interface InitBoostFormPayload {
  boostId: BoostEntity['id'];
  mode: 'stake' | 'unstake';

  // really, this should be separated
  walletAddress: string | undefined;
  balance: FetchAllBalancesResult;
  allowance: TokenAllowance[];

  // reducers below need to access the state
  state: BeefyState;
}

export const initiateBoostForm = createAsyncThunk<
  InitBoostFormPayload,
  InitBoostFormParams,
  { state: BeefyState }
>('boosts/initBoostForm', async ({ boostId, mode, walletAddress }, { getState }) => {
  const boost = selectBoostById(getState(), boostId);
  const vault = selectVaultById(getState(), boost.vaultId);
  const chain = selectChainById(getState(), boost.chainId);

  const balanceApi = await getBalanceApi(chain);

  const balanceRes: FetchAllBalancesResult = walletAddress
    ? await balanceApi.fetchAllBalances(getState(), [], [], [boost], walletAddress)
    : { tokens: [], boosts: [], govVaults: [] };

  const spenderAddress = boost.earnContractAddress;

  const allowanceApi = await getAllowanceApi(chain);
  const mooToken = selectErc20TokenByAddress(
    getState(),
    boost.chainId,
    vault.earnedTokenAddress,
    false
  );
  const allowanceRes =
    walletAddress && spenderAddress
      ? await allowanceApi.fetchTokensAllowance(
          getState(),
          [mooToken],
          walletAddress,
          spenderAddress
        )
      : [];

  return {
    boostId,
    mode,
    walletAddress,
    allowance: allowanceRes,
    balance: balanceRes,
    state: getState(),
  };
});
