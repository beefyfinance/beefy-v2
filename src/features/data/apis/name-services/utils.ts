import { uniq } from 'lodash-es';
import { type Address, getAddress, type Hash, isAddress, numberToHex } from 'viem';
import { namehash, normalize } from 'viem/ens';
import { ZERO_ADDRESS } from '../../../../helpers/addresses.ts';
import type { ChainId } from '../../entities/chain.ts';

export function hashDomain(domain: string): Hash {
  return namehash(domain);
}

export function normalizeDomain(domain: string): string {
  return normalize(domain);
}

export function normalizeAndHashDomain(domain: string): Hash {
  return hashDomain(normalizeDomain(domain));
}

export function normalizeAddress(address: unknown): Address | undefined {
  if (typeof address === 'string' && address.length === 42 && address !== ZERO_ADDRESS) {
    if (isAddress(address, { strict: false })) {
      return getAddress(address);
    }
  }

  return undefined;
}

export function getAllChainsFromTldToChain(tldToChain: Record<string, ChainId[]>): ChainId[] {
  return uniq(Object.values(tldToChain).flat());
}

/**
 * ENSIP-11: EVM compatible Chain Address Resolution
 * @see https://docs.ens.domains/ensip/11/
 **/
export function chainIdToCoinType(chainId: number): number {
  return (0x80000000 | chainId) >>> 0;
}

/**
 * ENSIP-19: Multichain Primary Names
 * @see https://docs.ens.domains/ensip/19/
 */
export function coinTypeToReverseDomain(coinType: number): string {
  switch (coinType) {
    case 60:
      return 'addr.reverse';
    case 0x8000_0000:
      return 'default.reverse';
    default: {
      if (coinType < 0 || coinType > 0xffff_ffff) {
        throw new Error(`Invalid coin type: ${coinType}`);
      }
      return `${numberToHex(coinType, { size: 4 }).slice(2)}.reverse`;
    }
  }
}

/**
 * ENSIP-19: Multichain Primary Names
 * @see https://docs.ens.domains/ensip/19/
 */
export function chainIdToReverseDomain(chainId: number): string {
  return coinTypeToReverseDomain(chainIdToCoinType(chainId));
}
