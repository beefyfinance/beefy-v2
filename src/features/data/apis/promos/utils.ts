export function extractChainId(path: string): string {
  const match = path.match(/.+\/chain\/(.+).json$/);
  const chainId = match?.[1];
  if (chainId) {
    return chainId;
  }
  throw new Error(`Could not extract chain id from path: ${path}`);
}
