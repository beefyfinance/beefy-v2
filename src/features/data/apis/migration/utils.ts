import { createAsyncThunk } from '@reduxjs/toolkit';
import type {
  CommonMigrationUpdateFulfilledPayload,
  MigratorExecuteProps,
  MigratorUpdateProps,
} from './migration-types';
import type { BeefyState } from '../../../../redux-types';
import { selectVaultById } from '../../selectors/vaults';
import { selectTokenByAddress } from '../../selectors/tokens';
import { selectUserBalanceToMigrateByVaultId } from '../../selectors/migration';
import type { Step } from '../../reducers/wallet/stepper';
import { walletActions } from '../../actions/wallet-actions';
import { startStepperWithSteps } from '../../actions/stepper';
import { isTokenErc20 } from '../../entities/token';
import { selectAllowanceByTokenAddress } from '../../selectors/allowances';
import type { VaultEntity } from '../../entities/vault';
import BigNumber from 'bignumber.js';
import type Web3 from 'web3';
import { selectChainById } from '../../selectors/chains';
import { getWalletConnectionApi, getWeb3Instance } from '../instances';

export function buildFetchBalance(
  id: string,
  fetchBalance: (
    vault: VaultEntity,
    web3: Web3,
    walletAddress: string,
    state: BeefyState
  ) => Promise<string>
) {
  return createAsyncThunk<
    CommonMigrationUpdateFulfilledPayload,
    MigratorUpdateProps,
    { state: BeefyState }
  >(`migration/${id}/update`, async ({ vaultId, walletAddress }, { getState }) => {
    const state = getState();
    const vault = selectVaultById(state, vaultId);
    const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
    const chain = selectChainById(state, vault.chainId);
    const web3 = await getWeb3Instance(chain);

    const balance = await fetchBalance(vault, web3, walletAddress, state);
    const fixedBalance = new BigNumber(balance).shiftedBy(-depositToken.decimals);
    console.debug(id, vault.id, fixedBalance.toNumber());
    return { vaultId, walletAddress, balance: fixedBalance, migrationId: id };
  });
}

export function buildExecute(
  id: string,
  unstakeCall: (
    vault: VaultEntity,
    web3: Web3,
    amount: BigNumber,
    state: BeefyState
    // eslint-disable-next-line
  ) => Promise<any>
) {
  return createAsyncThunk<void, MigratorExecuteProps, { state: BeefyState }>(
    `migration/${id}/execute`,
    async ({ vaultId, t, migrationId }, { getState, dispatch }) => {
      const steps: Step[] = [];
      const state = getState();
      const vault = selectVaultById(state, vaultId);
      const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
      const walletApi = await getWalletConnectionApi();
      const web3 = await walletApi.getConnectedWeb3Instance();
      const { balance } = selectUserBalanceToMigrateByVaultId(state, vaultId, migrationId);

      const call = await unstakeCall(vault, web3, balance, state);

      steps.push({
        step: 'migration',
        message: t('Vault-MigrationStart'),
        action: walletActions.migrateUnstake(
          call,
          vault,
          balance.shiftedBy(depositToken.decimals),
          migrationId
        ),
        pending: false,
        extraInfo: { vaultId },
      });

      if (isTokenErc20(depositToken)) {
        const allowance = selectAllowanceByTokenAddress(
          state,
          depositToken.chainId,
          depositToken.address,
          vault.earnContractAddress
        );
        if (allowance.lt(balance)) {
          steps.push({
            step: 'approve',
            message: t('Vault-ApproveMsg'),
            action: walletActions.approval(depositToken, vault.earnContractAddress),
            pending: false,
          });
        }
      }

      steps.push({
        step: 'deposit',
        message: t('Vault-TxnConfirm', { type: t('Deposit-noun') }),
        action: walletActions.deposit(vault, balance, true),
        pending: false,
        extraInfo: { vaultId: vault.id },
      });

      dispatch(startStepperWithSteps(steps, vault.chainId));
    }
  );
}
