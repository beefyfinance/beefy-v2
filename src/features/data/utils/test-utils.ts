import { Action } from 'redux';
import { configureStore } from '@reduxjs/toolkit';
import { fetchAllBoosts } from '../actions/boosts';
import { fetchChainConfigs } from '../actions/chains';
import { fetchAllPricesAction } from '../actions/prices';
import { fetchAllVaults } from '../actions/vaults';
import { selectAllChains } from '../selectors/chains';
import mockPrices from './mock-prices.json';
import mockLPPrices from './mock-lp-prices.json';
import { rootReducer } from '../reducers';

/**
 * Create a new BeefyState store with some data included
 * This is used in tests to get a decent starting state
 * when doing things like TVL or APY computation
 * TODO: maybe memoize that, we have to deep freeze the output for that
 */
export async function getBeefyTestingStore() {
  // here, the store has all the proper TS typings
  const store = configureStore({
    reducer: rootReducer,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        // because we use BigNumber which is not serializable by default
        // we disable rerialization altogether
        // a better solution would be to allow serialization of the store
        serializableCheck: false,

        // this makes the old code bug
        immutableCheck: false,
      }),
  });

  let state = store.getState();

  // trigger all configs actions
  await store.dispatch(fetchChainConfigs());
  state = store.getState();
  const chains = selectAllChains(state);
  await store.dispatch(fetchAllVaults({}));
  await store.dispatch(fetchAllBoosts());

  // mock token prices
  store.dispatch({
    type: fetchAllPricesAction.fulfilled,
    payload: { ...mockPrices, ...mockLPPrices },
  });

  return store;
}
