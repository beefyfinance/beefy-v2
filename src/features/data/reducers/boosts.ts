import { createSlice } from '@reduxjs/toolkit';
import { WritableDraft } from 'immer/dist/internal';
import { fetchAllBoosts } from '../actions/boosts';
import { BoostConfig } from '../apis/config';
import { BoostEntity, isBoostActive, isBoostPreStake } from '../entities/boost';
import { ChainEntity } from '../entities/chain';
import { VaultEntity } from '../entities/vault';
import { NormalizedEntity } from '../utils/normalized-entity';

/**
 * State containing Vault infos
 */
export type BoostsState = NormalizedEntity<BoostEntity> & {
  byVaultId: {
    [vaultId: VaultEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
  byChainId: {
    [chainId: ChainEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
};
export const initialBoostsState: BoostsState = {
  byId: {},
  allIds: [],
  byVaultId: {},
  byChainId: {},
};

export const boostsSlice = createSlice({
  name: 'boosts',
  initialState: initialBoostsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // when boost list is fetched, add all new tokens
    builder.addCase(fetchAllBoosts.fulfilled, (sliceState, action) => {
      for (const [chainId, boosts] of Object.entries(action.payload)) {
        for (const boost of boosts) {
          addBoostToState(sliceState, chainId, boost);
        }
      }
    });
  },
});

function addBoostToState(
  sliceState: WritableDraft<BoostsState>,
  chainId: ChainEntity['id'],
  apiBoost: BoostConfig
) {
  if (apiBoost.id in sliceState.byId) {
    return;
  }

  const boost: BoostEntity = {
    id: apiBoost.id,
    chainId: chainId,
    assets: apiBoost.assets,
    earnedTokenId: apiBoost.earnedOracleId,
    earnContractAddress: apiBoost.earnContractAddress,
    logo: apiBoost.logo,
    name: apiBoost.name,
    partnerIds: apiBoost.partners ? apiBoost.partners.map(p => p.website) : [],
    status: apiBoost.status as BoostEntity['status'],
    vaultId: apiBoost.poolId,
  };
  sliceState.byId[boost.id] = boost;
  sliceState.allIds.push(boost.id);

  // add to vault id index
  if (sliceState.byVaultId[boost.vaultId] === undefined) {
    sliceState.byVaultId[boost.vaultId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byVaultId[boost.vaultId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byVaultId[boost.vaultId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byVaultId[boost.vaultId].prestakeBoostsIds.push(boost.id);
  }

  // add to chain id index
  if (sliceState.byChainId[chainId] === undefined) {
    sliceState.byChainId[chainId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byChainId[chainId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byChainId[chainId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byChainId[chainId].prestakeBoostsIds.push(boost.id);
  }
}
