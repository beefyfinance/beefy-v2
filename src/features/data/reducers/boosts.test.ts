import BigNumber from 'bignumber.js';
import { fetchAllBoosts, FulfilledAllBoostsPayload } from '../actions/boosts';
import {
  fetchAllContractDataByChainAction,
  FetchAllContractDataFulfilledPayload,
} from '../actions/contract-data';
import { getBeefyTestingStore } from '../utils/test-utils';
import { boostsSlice, initialBoostsState, recomputeBoostStatus } from './boosts';

describe('Boosts slice tests', () => {
  it('should update state on fulfilled boosts list', () => {
    const payload: FulfilledAllBoostsPayload = {
      harmony: [
        // one with a partnership
        {
          id: 'moo_banana-pera',
          poolId: 'banana-banana-eol',
          name: 'Pera',
          logo: 'degens/BANANA.svg',
          earnedToken: 'PERA',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xb9D8592E16A9c1a3AE6021CDDb324EaC1Cbc70d6',
          earnContractAddress: '0x38247fCE28480A7BEF2CB7aD134ce091Bd2E1a82',
          earnedOracle: 'tokens',
          earnedOracleId: 'PERA',
          partnership: true,
          status: 'active',
          isMooStaked: true,
          partners: [
            {
              logo: 'stake/pera/logo.png',
              background: 'stake/pera/bg.png',
              text: "Pera Finance is a decentralized trading competition platform with multi-layered yield farming opportunities. Traders, liquidity providers and holders yield farm together through the DeFi's first decentralized trading competition. Every on-chain PERA transaction (transfer, trade, or liquidity addition/ removal) generates a 2 % transaction fee.",
              website: 'https://pera.finance/',
              social: {
                telegram: 'https://t.me/perafinance',
                twitter: 'https://twitter.com/perafinance',
              },
            },
          ],
        },

        // one closed boost
        {
          id: 'moo_aave-btc',
          poolId: 'aave-wbtc',
          name: 'Iron / Garuda / Fanatics',
          logo: 'single-assets/BTCB.svg',
          earnedToken: 'mooPolygonBIFI',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xfEcf784F48125ccb7d8855cdda7C5ED6b5024Cb3',
          earnContractAddress: '0x20948Cad130c3D7B24d27CC66163b4aaed4684F0',
          earnedOracle: 'tokens',
          earnedOracleId: 'BIFI',
          partnership: true,
          status: 'closed',
          isMooStaked: true,
          partners: [],
        },

        // one with pre-staking boost
        {
          id: 'moo_aave-eth',
          poolId: 'aave-eth',
          name: 'Iron / Garuda / Fanatics',
          logo: 'single-assets/ETH.svg',
          earnedToken: 'mooPolygonBIFI',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xfEcf784F48125ccb7d8855cdda7C5ED6b5024Cb3',
          earnContractAddress: '0x9B508ad657ed5A139D1a7c97fD84d7B7240849Cf',
          earnedOracle: 'tokens',
          earnedOracleId: 'BIFI',
          partnership: true,
          status: 'prestake',
          isMooStaked: true,
        },
      ],
    };
    const action = { type: fetchAllBoosts.fulfilled, payload: payload };
    const state = boostsSlice.reducer(initialBoostsState, action);
    expect(state).toMatchSnapshot();

    // getting the same vaults don't update the state object
    const beforeReDispatch = Object.values(state.byId)[0];
    const newState = boostsSlice.reducer(state, action);
    const afterReDispatch = newState.byId[beforeReDispatch.id];
    expect(beforeReDispatch).toBe(afterReDispatch);
  });

  it('should update state on fulfilled boost contract data', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const state = store.getState();

    // given we have received data for some boosts
    const initPayload: FulfilledAllBoostsPayload = {
      bsc: [
        {
          id: 'moo_banana-banana-busd-bitcrush',
          poolId: 'aave-wbtc',
          name: 'Iron / Garuda / Fanatics',
          logo: 'single-assets/BTCB.svg',
          earnedToken: 'mooPolygonBIFI',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xfEcf784F48125ccb7d8855cdda7C5ED6b5024Cb3',
          earnContractAddress: '0x20948Cad130c3D7B24d27CC66163b4aaed4684F0',
          earnedOracle: 'tokens',
          earnedOracleId: 'BIFI',
          partnership: true,
          status: '',
          isMooStaked: true,
          partners: [],
        },
        {
          id: 'moo_banana-bnb-stars-mogul2',
          poolId: 'aave-eth',
          name: 'Iron / Garuda / Fanatics',
          logo: 'single-assets/ETH.svg',
          earnedToken: 'mooPolygonBIFI',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xfEcf784F48125ccb7d8855cdda7C5ED6b5024Cb3',
          earnContractAddress: '0x9B508ad657ed5A139D1a7c97fD84d7B7240849Cf',
          earnedOracle: 'tokens',
          earnedOracleId: 'BIFI',
          partnership: true,
          status: '',
          isMooStaked: true,
        },
      ],
    };
    const initAction = { type: fetchAllBoosts.fulfilled, payload: initPayload };
    const initState = boostsSlice.reducer(initialBoostsState, initAction);

    // We want to make sure we handle the action properly
    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        govVaults: [],
        standardVaults: [],
        boosts: [
          {
            id: 'moo_banana-banana-busd-bitcrush',
            periodFinish: new Date(Date.UTC(2022, 0, 1, 0, 0, 0)), // expired boost
            rewardRate: new BigNumber(0.4),
            totalSupply: new BigNumber(12345),
          },
          {
            id: 'moo_banana-bnb-stars-mogul2',
            periodFinish: new Date(Date.UTC(2250, 0, 1, 0, 0, 0)), // active boost
            rewardRate: new BigNumber(0.4),
            totalSupply: new BigNumber(12345),
          },
          {
            id: 'moo_banana-crush-wbnb-bitcrush',
            periodFinish: null, // prestake boost
            rewardRate: new BigNumber(0.4),
            totalSupply: new BigNumber(12345),
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const newState = boostsSlice.reducer(initState, action);
    expect({ ...newState, periodFinish: 'tested-below' }).toMatchSnapshot();
    // team all around the world, make sure we test utc snapshots
    expect(
      Object.entries(newState.periodFinish).map(([boostId, date]) => [
        boostId,
        date ? date.toUTCString() : date,
      ])
    ).toMatchSnapshot();

    // We don't want the state to change reference when recomputing boost status
    const stateAfterRecompute1 = boostsSlice.reducer(newState, action);
    expect(stateAfterRecompute1).toBe(newState);

    // We don't want the state to change reference when recomputing boost status
    const stateAfterRecompute2 = boostsSlice.reducer(stateAfterRecompute1, {
      type: recomputeBoostStatus,
    });
    expect(stateAfterRecompute2).toBe(newState);
  });

  it('should correctly handle mooFuse boost config', () => {
    const payload: FulfilledAllBoostsPayload = {
      fuse: [
        {
          id: 'moo_fusefi-wfuse-usdc-fuse',
          poolId: 'voltage-wfuse-usdc',
          name: 'Fuse',
          assets: ['USDC', 'FUSE'],
          earnedToken: 'mooFuse',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0x2C43DBef81ABa6b95799FD2aEc738Cd721ba77f3',
          earnContractAddress: '0x405EE7F4f067604b787346bC22ACb66b06b15A4B',
          earnedOracle: 'tokens',
          earnedOracleId: 'WFUSE',
          partnership: true,
          status: 'active',
          isMooStaked: true,
          partners: [],
        },
      ],
    };
    const action = { type: fetchAllBoosts.fulfilled, payload: payload };
    const state = boostsSlice.reducer(initialBoostsState, action);
    expect(state).toMatchSnapshot();
  });

  it('should correctly handle wfuse-elon boost config', () => {
    const payload: FulfilledAllBoostsPayload = {
      fuse: [
        {
          id: 'moo_voltage-wfuse-elon-elon',
          poolId: 'voltage-wfuse-elon',
          name: 'Dogelon Mars x Fuse',
          assets: ['ELON', 'FUSE'],
          earnedToken: 'ELON-FUSE LP',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xe418c323fA450e7e18c4dB304bEFC7ffF92D2Cc1',
          earnContractAddress: '0xc3a4fdcba79DB04b4C3e352b1C467B3Ba909D84A',
          earnedOracle: 'lps',
          earnedOracleId: 'voltage-wfuse-elon',
          partnership: true,
          status: 'active',
          isMooStaked: true,
          partners: [],
        },
      ],
    };
    const action = { type: fetchAllBoosts.fulfilled, payload: payload };
    const state = boostsSlice.reducer(initialBoostsState, action);
    expect(state).toMatchSnapshot();
  });
});
