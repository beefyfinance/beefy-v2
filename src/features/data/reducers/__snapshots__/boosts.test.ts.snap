// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boosts slice tests should correctly handle mooFuse boost config 1`] = `
Object {
  "allIds": Array [
    "moo_fusefi-wfuse-usdc-fuse",
  ],
  "byChainId": Object {
    "fuse": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_fusefi-wfuse-usdc-fuse",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
  },
  "byId": Object {
    "moo_fusefi-wfuse-usdc-fuse": Object {
      "assets": Array [
        "USDC",
        "FUSE",
      ],
      "chainId": "fuse",
      "earnContractAddress": "0x405EE7F4f067604b787346bC22ACb66b06b15A4B",
      "earnedTokenId": "mooFuse",
      "id": "moo_fusefi-wfuse-usdc-fuse",
      "logo": undefined,
      "name": "Fuse",
      "partnerIds": Array [],
      "vaultId": "voltage-wfuse-usdc",
    },
  },
  "byVaultId": Object {
    "voltage-wfuse-usdc": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_fusefi-wfuse-usdc-fuse",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
  },
  "periodFinish": Object {},
}
`;

exports[`Boosts slice tests should correctly handle wfuse-elon boost config 1`] = `
Object {
  "allIds": Array [
    "moo_voltage-wfuse-elon-elon",
  ],
  "byChainId": Object {
    "fuse": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_voltage-wfuse-elon-elon",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
  },
  "byId": Object {
    "moo_voltage-wfuse-elon-elon": Object {
      "assets": Array [
        "ELON",
        "FUSE",
      ],
      "chainId": "fuse",
      "earnContractAddress": "0xc3a4fdcba79DB04b4C3e352b1C467B3Ba909D84A",
      "earnedTokenId": "voltage-wfuse-elon",
      "id": "moo_voltage-wfuse-elon-elon",
      "logo": undefined,
      "name": "Dogelon Mars x Fuse",
      "partnerIds": Array [],
      "vaultId": "voltage-wfuse-elon",
    },
  },
  "byVaultId": Object {
    "voltage-wfuse-elon": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_voltage-wfuse-elon-elon",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
  },
  "periodFinish": Object {},
}
`;

exports[`Boosts slice tests should update state on fulfilled boost contract data 1`] = `
Object {
  "allIds": Array [
    "moo_banana-banana-busd-bitcrush",
    "moo_banana-bnb-stars-mogul2",
  ],
  "byChainId": Object {
    "bsc": Object {
      "activeBoostsIds": Array [
        "moo_banana-bnb-stars-mogul2",
      ],
      "allBoostsIds": Array [
        "moo_banana-banana-busd-bitcrush",
        "moo_banana-bnb-stars-mogul2",
      ],
      "expiredBoostsIds": Array [
        "moo_banana-banana-busd-bitcrush",
      ],
      "prestakeBoostsIds": Array [],
    },
  },
  "byId": Object {
    "moo_banana-banana-busd-bitcrush": Object {
      "assets": undefined,
      "chainId": "bsc",
      "earnContractAddress": "0x20948Cad130c3D7B24d27CC66163b4aaed4684F0",
      "earnedTokenId": "mooPolygonBIFI",
      "id": "moo_banana-banana-busd-bitcrush",
      "logo": "single-assets/BTCB.svg",
      "name": "Iron / Garuda / Fanatics",
      "partnerIds": Array [],
      "vaultId": "aave-wbtc",
    },
    "moo_banana-bnb-stars-mogul2": Object {
      "assets": undefined,
      "chainId": "bsc",
      "earnContractAddress": "0x9B508ad657ed5A139D1a7c97fD84d7B7240849Cf",
      "earnedTokenId": "mooPolygonBIFI",
      "id": "moo_banana-bnb-stars-mogul2",
      "logo": "single-assets/ETH.svg",
      "name": "Iron / Garuda / Fanatics",
      "partnerIds": Array [],
      "vaultId": "aave-eth",
    },
  },
  "byVaultId": Object {
    "aave-eth": Object {
      "activeBoostsIds": Array [
        "moo_banana-bnb-stars-mogul2",
      ],
      "allBoostsIds": Array [
        "moo_banana-bnb-stars-mogul2",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
    "aave-wbtc": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_banana-banana-busd-bitcrush",
      ],
      "expiredBoostsIds": Array [
        "moo_banana-banana-busd-bitcrush",
      ],
      "prestakeBoostsIds": Array [],
    },
  },
  "periodFinish": "tested-below",
}
`;

exports[`Boosts slice tests should update state on fulfilled boost contract data 2`] = `
Array [
  Array [
    "moo_banana-banana-busd-bitcrush",
    "Sat, 01 Jan 2022 00:00:00 GMT",
  ],
  Array [
    "moo_banana-bnb-stars-mogul2",
    "Tue, 01 Jan 2250 00:00:00 GMT",
  ],
  Array [
    "moo_banana-crush-wbnb-bitcrush",
    null,
  ],
]
`;

exports[`Boosts slice tests should update state on fulfilled boosts list 1`] = `
Object {
  "allIds": Array [
    "moo_banana-pera",
    "moo_aave-btc",
    "moo_aave-eth",
  ],
  "byChainId": Object {
    "harmony": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_banana-pera",
        "moo_aave-btc",
        "moo_aave-eth",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
  },
  "byId": Object {
    "moo_aave-btc": Object {
      "assets": undefined,
      "chainId": "harmony",
      "earnContractAddress": "0x20948Cad130c3D7B24d27CC66163b4aaed4684F0",
      "earnedTokenId": "mooPolygonBIFI",
      "id": "moo_aave-btc",
      "logo": "single-assets/BTCB.svg",
      "name": "Iron / Garuda / Fanatics",
      "partnerIds": Array [],
      "vaultId": "aave-wbtc",
    },
    "moo_aave-eth": Object {
      "assets": undefined,
      "chainId": "harmony",
      "earnContractAddress": "0x9B508ad657ed5A139D1a7c97fD84d7B7240849Cf",
      "earnedTokenId": "mooPolygonBIFI",
      "id": "moo_aave-eth",
      "logo": "single-assets/ETH.svg",
      "name": "Iron / Garuda / Fanatics",
      "partnerIds": Array [],
      "vaultId": "aave-eth",
    },
    "moo_banana-pera": Object {
      "assets": undefined,
      "chainId": "harmony",
      "earnContractAddress": "0x38247fCE28480A7BEF2CB7aD134ce091Bd2E1a82",
      "earnedTokenId": "PERA",
      "id": "moo_banana-pera",
      "logo": "degens/BANANA.svg",
      "name": "Pera",
      "partnerIds": Array [
        "https://pera.finance/",
      ],
      "vaultId": "banana-banana-eol",
    },
  },
  "byVaultId": Object {
    "aave-eth": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_aave-eth",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
    "aave-wbtc": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_aave-btc",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
    "banana-banana-eol": Object {
      "activeBoostsIds": Array [],
      "allBoostsIds": Array [
        "moo_banana-pera",
      ],
      "expiredBoostsIds": Array [],
      "prestakeBoostsIds": Array [],
    },
  },
  "periodFinish": Object {},
}
`;
